### PR TITLE
muxassign assign.json - remove whitespace from keys

### PR DESCRIPTION
**Pull request by Matsyir**
Pretty much as title says - remove whitespace from keys in assign.json (and subsequent duplicates). By also removing whitespace from the original folder name we're checking, we should be able to include tons more alternate spellings, while also making the assign.json file easier to populate.

My other branch [assign-json-testcases](https://github.com/Matsyir/internal/tree/assign-json-testcases/assign-json-testcases) includes a very basic nodejs script to compare this new name-matching system with the old one, to confirm this doesn't result in less matches/less functionality. It's pretty much just 1 nodejs script + the json so you shouldn't need to download the whole fork if you want to run it. The script used to generate this assign.json file is also in that test file, commented. (It was then formatted using vscode)

Note that the nodejs script is not exactly 1:1 using the old "intended" values for the tests, for a few outliers, because I noticed some bugs in the original assign.json. For example, not all keys were already lowercased, some values had leading or trailing whitespaces (e.g "arcade.ini "), and some had malformed extensions (e.g "arcade.in"). After manually fixing a few of these edge cases, the test cases seem to indicate there's no loss of results by removing whitespace like this.

This is combined with [another PR on frontend](https://github.com/MustardOS/frontend/pull/6), so that it strips whitespaces from the folder names that will be matched with these keys.

### CODE REVIEW

**Comment by Matsyir** [comment]
I thought about this a bit more, and figured, why not also remove dashes and underscores if we're doing spaces? I also considered entirely removing any non-alphanumeric characters, but that only saves 2 entries over removing space+dash+underscore, plus I didn't see a regex replace among the string functions, so I think just space+dash+underscore is efficient/inclusive enough.

I just created a new test script for that (found at the same link above), and it does seem to work well, with even less keys required once again.
Current/original assign.json: 707 keys
Whitespace removed: 603 keys
Whitespace & -_ removed: 456 keys
Only conserving alphanumeric: 454 keys (g+w & mr.boom)

However, there's 1 false positive, but from a quick glance through the list, it seems like that might've been another bug with the original assign.json, but I'm not sure. Here is a screenshot of the false positive, from the original/current assign.json:
Code_9nyUPbypLN

After we exclude "-_", then "neogeo": "snk neo geo cd.ini" gets overwritten by "neo-geo": "snk neo geo.ini".
I mean... if "neo-geo" and all those similar variations are supposed to be "snk neo geo.ini", then I'd think "neogeo" should be too? It's probably not supposed to have cd.ini? Or am I wrong on that? Don't know much about the neogeo consoles so this is my best guess based on the formatting involved.
Also, "neogeo" is up along with the other "snk neo geo.ini"s, rather than being below along with the other "snk neo geo cd.ini"s, if that makes sense.

Basically, I think it'd also help to extend this PR to remove dashes & underscores from keys, on top of spaces. I'll see what you think before putting in any more commits.

The changes required to also remove dashes & underscores are pretty small, basically just copying the new-assign2.json from my test branch, and then removing those before the json 'search' in C (either adding 2 more str_remchar calls or using some kind of multi-char replace if available).

**Comment by Matsyir** [comment]
> This is removing readability. I think I'm missing the justification of this pull request?

It would make this process/the assign.json file automatically detect tons more variations of folder names, so it can automatically assign cores to various "new" folder names, without us having to specifically define those name variations, so would also allow us to more easily include variations when adding to the list, even unintentional ones.

It also should have a marginal performance benefit of making the assign.json more compact/dense, making it quicker to "search" through and find a matching key, though that wasn't really a factor in suggesting this change. E.g, if we take the tests where it removes space+dash+underscore, we get the same functionality out of an assign.json with only 456 keys than we do the current one with 707 keys (well, assuming that 1 test case isn't problematic).

Here's one example that didn't work for me, that would work after this change:
My Folder name: Nintendo - Game Boy Advance (not in the list)
Closest in the existing assign.json list: nintendo - gameboy advance

Didn't match because there was no variation defined with a space in the word "gameboy". That's just one example, but it'd be possible to make hundreds of examples of new folder names that would be included with this change.

I think just space alone is the most important to include here, but as I said I think it'd make sense to do dash+underscore as well.

**Comment by xonglebongle** [comment]
Ah okay I see what you mean.  This will have to be thoroughly tested first. 🤘

**Comment by Matsyir** [comment]
Yeah, totally agree, that's why I made sure to write at least some basic tests before making the PR(s).

To summarize [those tests](https://github.com/Matsyir/internal/tree/assign-json-testcases/assign-json-testcases)' intentions: they basically use the original assign.json's keys as a list of "test folder names", to process and compare against the new assign.json the same way muOS will. So, lowercasing the folder name, then removing whitespace, then comparing with the new assign.json - or removing whitespace+dash+underscore depending which test. In other words, any folder name that matched the old keys should match the new ones with these changes.

Of course, it'll still need to be tested on the device itself like anything else. But it seems like a relatively safe change to me, the C involved seems pretty inconsequential (especially considering there was already string processing on the folder name - tolower), and based on the test scripts, the assign.json seems safely inclusive of all previously existing keys.

I'm curious what you think of the failed test case? Would you agree that the `neogeo` key is likely meant to be `snk neo geo.ini` rather than `snk neo geo cd.ini`? If it's supposed to be cd, it could be problematic, would either require an exception-case (ew no) or sticking to removing space only rather than space+dash+underscore.

(space has 0 failed test cases[after fixing a few things in the original like "arcade.in" & "arcade.ini ", uppercase/untrimmed keys], space+dash+underscore just that neogeo one)

**Comment by xonglebongle** [comment]
The neogeo case is most likely just a human error 😅

**Comment by Matsyir** [comment]
That'd be perfect, and totally understandable for such a long list! Haha, was just hoping it wasn't some odd exception where `neogeo` as 1 word commonly means something different than with dashes/spaces or something like that. Just googled it a bit more and yeah, doesn't seem to be the case so it should be good.

In that case, we're likely able to safely remove space+dash+underscore, using the existing [new-assign2.json](https://github.com/Matsyir/internal/blob/assign-json-testcases/assign-json-testcases/new-assign2.json) from my test branch (which fixes the neogeo case and sets it to `snk neo geo.ini` rather than CD).

If you want me to update the PRs for space+dash+underscore, let me know if simply adding 2 more `str_remchar` calls for the frontend part is OK - if so I can slap it in and update the json. The existing PRs are only removing whitespace.

If we prefer to do a custom tolower+multi-char-replace function for efficiency (not sure if this is called in a looping context), it's probably best if someone else does that, since I don't _really_ know C, as I mentioned.
I mean I'm mostly a C#/Java guy, and do understand pointers (not really their C syntax tho), so not _completely_ lost, but would take me way too long for what it is... especially considering it's processing _strings_ in C of all things.
